src/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -316,32 +312,12 @@ internal static unsafe string EscapeUnescapeIri(char* pInput, int start, int end
 
                     if (bufferRemaining < MaxNumberOfBytesEncoded * percentEncodingLen)
                     {
-                        int newBufferLength = 0;
-
                         checked
                         {
-                            // may need more memory since we didn't anticipate escaping
-                            newBufferLength = dest.Length + bufferCapacityIncrease;
                             bufferRemaining += bufferCapacityIncrease;
                         }
 
-                        char[] newDest = new char[newBufferLength];
-
-                        fixed (char* pNewDest = newDest)
-                        {
-                            Buffer.MemoryCopy((byte*)pDest, (byte*)pNewDest, newBufferLength * sizeof(char), destOffset * sizeof(char));
-                        }
-
-                        if (destHandle.IsAllocated)
-                        {
-                            destHandle.Free();
-                        }
-
-                        dest = newDest;
-
-                        // re-pin new dest[] array
-                        destHandle = GCHandle.Alloc(dest, GCHandleType.Pinned);
-                        pDest = (char*)destHandle.AddrOfPinnedObject();
+                        dest.AppendDefault(bufferCapacityIncrease);
```

Comment:
As noted earlier, this doesn't need a new helper on ValueStringBuilder.  That type is compiled into a bunch of different places: we only want functionality on it that is really universal; everything else can be done by the consumers.  You can just build this on top of ValueStringBuilder's public APIs.

Reply:
Methods are extracted out of ValueStringBuilder. Changes reverted.